icu4j/main/classes/core/src/com/ibm/icu/impl/ICUResourceBundle.java

```diff
@@ -1261,8 +1263,90 @@ private static boolean localeIDStartsWithLangSubtag(String localeID, String lang
+            return pair1[0].compareTo(pair2[0]);
+        }
+    };
+
+    private static String getExplicitParent(String localeID) {
+        String[] temp = { localeID, "" };
+        int parentIndex = Arrays.binarySearch(LocaleFallbackTables.PARENT_LOCALE_TABLE, temp,COMPARE_FIRST_ELEMENT);
+        return (parentIndex < 0) ? null : LocaleFallbackTables.PARENT_LOCALE_TABLE[parentIndex][1];
+    }
+
+    private static String getDefaultScript(String language, String region) {
+        String[] temp = { language + "_" + region, "" };
+        int scriptIndex = Arrays.binarySearch(LocaleFallbackTables.DEFAULT_SCRIPT_TABLE, temp, COMPARE_FIRST_ELEMENT);
+        if (scriptIndex < 0) {
+            temp[0] = language;
+            scriptIndex = Arrays.binarySearch(LocaleFallbackTables.DEFAULT_SCRIPT_TABLE, temp, COMPARE_FIRST_ELEMENT);
+        }
+        return (scriptIndex < 0) ? "Latn" : LocaleFallbackTables.DEFAULT_SCRIPT_TABLE[scriptIndex][1];
+    }
+
+    private static String getParentLocaleID(String name, String origName) {
+        // early out if the locale ID has a variant code or ends with _
+        if (name.endsWith("_") || !ULocale.getVariant(name).isEmpty()) {
+            int lastUnderbarPos = name.lastIndexOf('_');
+            if (lastUnderbarPos >= 0) {
+                return name.substring(0, lastUnderbarPos);
+            } else {
+                return null;
+            }
+        }
+
+        ULocale nameLocale = new ULocale(name);
+        String language = nameLocale.getLanguage();
+        String script = nameLocale.getScript();
+        String region = nameLocale.getCountry();
+
+        // first look the locale ID up in the parent locale table; if that table specifies a parent
+        // for it, return that
+        String parentID = getExplicitParent(name);
+        if (parentID != null) {
+            return parentID.equals("root") ? null : parentID;
+        }
+
+        // if it's not in the parent locale table, figure out the fallback script algorithmically
+        // (see CLDR-15265 for an explanation of the algorithm)
+        if (!script.isEmpty() && !region.isEmpty()) {
+            // if "name" has both script and region, is the script the default script?
+            // - if so, remove it and keep the region
+            // - if not, remove the region and keep the script
+            if (getDefaultScript(language, region).equals(script)) {
+                return language + "_" + region;
+            } else {
+                return language + "_" + script;
+            }
+        } else if (!region.isEmpty()) {
+            // if "name" has region but not script, did the original locale ID specify a script?
+            // - if yes, replace the region with the script from the original locale ID
+            // - if no, replace the region with the default script for that language and region
+            String origNameScript = ULocale.getScript(origName);
```

Comment:
It looks like there is a fair bit of conversion back & forth between String and ULocale.
Maybe add a TODO to optimize this later, such as using a struct like class LSR in the LocaleMatcher?

Reply:
The idea originally was to keep it all in the String domain, but the easiest way I could find to break the string up into language, script, region, and variant.  I think I considered LSR, but it didn't have that functionality.  If you've got a better way of doing this, I'm all ears.  I did go ahead and add a TODO.